File: smallest_subarray.py
#https://onlinejudge.org/index.php?option=onlinejudge&Itemid=8&page=show_problem&problem=2531
c=[0]*100000
for t in range(int(input())):
#for t in range(1):
    n,m,k=map(int,input().split())
    #n,m,k=20,12,4
    a=[1,2,3]
    cnt=0
    mi=1000000000000
    queue=[]
    for i in range(3,n):
        sum=a[i-1]+a[i-2]+a[i-3]
        sum%=m
        a.append(sum+1)
    for i in range(n):
        if a[i]>=1 and a[i]<=k:
            if c[a[i]-1]==0:
                cnt+=1
            c[a[i]-1]+=1
            queue.append(i)
        while cnt==k:
            temp=i-queue[0]+1
            mi=min(mi,temp)
            c[a[queue[0]]-1]-=1
            if c[a[queue[0]]-1]==0:
                cnt-=1
            queue.pop(0)
    print(a)
    if mi==1000000000000:
        print("Case #{}: {}".format(t+1,"sequence nai"))
    else:
        print("Case #{}: {}".format(t+1,mi))

File: smallest_subarray.cpp
//https://onlinejudge.org/index.php?option=onlinejudge&Itemid=8&page=show_problem&problem=2531
#include<iostream>
#include<cstdio>
#include<cstring>
#include<utility>
#include<set>
#include<queue>
using namespace std;
 
typedef long long int64;
typedef pair<int,int>pii;
 
const int INF = 0x3f3f3f3f;
const int MAXN = 1000010;
 
int n, m, k;
int arr[MAXN];
int cnt[MAXN];
 
void init(){
    arr[1] = 1;
    arr[2] = 2;
    arr[3] = 3;
    for(int i=4; i<=n; ++i)
        arr[i] = (arr[i-1]+arr[i-2]+arr[i-3])%m + 1;
	memset(cnt, 0, sizeof(cnt));
}
 
int main(){
    int nCase, cas=1;
    scanf("%d", &nCase);
 
    while(nCase--){
        
        scanf("%d%d%d", &n,&m,&k);
 
		init();
		int minx = INF;
        int count=0;
		queue<int>Q;
 
		for(int i=1; i<=n; ++i){
		    if(arr[i]>=1 && arr[i]<=k){
		        Q.push(i);
                if(cnt[arr[i]]++ == 0){
                    ++count;
 
                }
		        while(count == k){
		            int tmp = i - Q.front() + 1;
		            minx = min(tmp, minx);
		            int val = arr[Q.front()];
		            if(--cnt[val]==0){
                        --count;
		            }
		            Q.pop();
 
		        }
		    }
		}
        printf("Case %d: ", cas++);
		if(minx == INF) puts("sequence nai");
		else printf("%d\n", minx);
    }
    return 0; 
}